driver.go

```diff
@@ -126,14 +125,111 @@ func (d MySQLDriver) Open(dsn string) (driver.Conn, error) {
 		maxap, err := mc.getSystemVar("max_allowed_packet")
 		if err != nil {
 			mc.Close()
-			return nil, err
+			return err
 		}
 		mc.maxAllowedPacket = stringToInt(maxap) - 1
 	}
 	if mc.maxAllowedPacket < maxPacketSize {
 		mc.maxWriteSize = mc.maxAllowedPacket
 	}
 
+	return err
+}
+
+//Connect opens a new connection without using a DSN
+func (c MySQLConnector) Connect(cxt context.Context) (driver.Conn, error) {
+	var err error
+
+	//Validate the connection parameters
+	//the following are required User,Pass,Net,Addr,DBName
+	//Pass may be blank
+	//The other optional parameters are not checks
+	//as GO will automatically enforce proper bool types on the options
+	if len(c.Cfg.User) > 32 || len(c.Cfg.User) <= 0 {
+		return nil, errInvalidUser
+	}
+
+	if len(c.Cfg.Addr) <= 0 {
+		return nil, errInvalidAddr
+	}
+
+	if len(c.Cfg.DBName) <= 0 {
+		return nil, errInvalidDBName
+	}
+
+	if c.Cfg.Net != "tcp" {
+		return nil, errInvalidNet
+	}
+
+	//New mysqlConn
+	mc := &mysqlConn{
+		maxAllowedPacket: maxPacketSize,
+		maxWriteSize:     maxPacketSize - 1,
+		closech:          make(chan struct{}),
+		cfg:              c.Cfg,
+		parseTime:        c.Cfg.ParseTime,
+	}
+
+	//Check if the there is a canelation before creating the connection
+	select {
+	case <-cxt.Done():
+		return nil, cxt.Err()
+	default:
+		//Connect to the server and setting the connection settings
+		err = connectServer(mc)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	//Check to see if there was a canelation during creating the connection
+	select {
+	case <-cxt.Done():
+		err = mc.Close()
```

Comment:
Checking `ctx.Done()` after connectServer() is really required?

Reply:
We do not have to, I debated just doing it once before the connection or just doing it after the connection. Once before is probably the correct/best option. The question really is at what points to do you want to check if there has been a cancelation or deadline hit that this is no longer needed. Ideally you exit as soon as possible as fast as possible. Checking pre-connection is probably adequate.